sympy/physics/quantum/qapply.py

```diff
@@ -1,206 +1,757 @@
+#
+# Moreover:
+#  5. do not use expand() indiscriminately: Only do an expansion of Add if required to apply operators, so
+#     keeping expansion to the bare minimum. As default, distribute commutative factors over sums as this allows
+#     terms to cancel out thus simplifying the expression. In order to simplify commutative factors qapply()
+#     employs the three mayor hint functions of expand() Mul._eval_expand_mul(), Pow._eval_expand_power_base()
+#     and Pow._eval_expand_power_exp(). See options mul, power_base and power_exp.
+#  6. apply recursion along the structure of the expression but don't use recursion for iterations:
+#     this avoids hitting the Python recursion limit in very complex expressions or powers with
+#     large exponents
+#  7. No need to try to maintain subclass types of objects and expressions, as we simply cannot know whether
+#     a product is still in a subclass of Mul etc. So use the top types.
+
+# Most rules might at a later time be implemented as methods of the object classes, however I decided
+# to be minimally invasive into the existing code base and to implement each set of rules locally
+# in this file, so this file is a 1:1 replacement of the previous version of qapply.py
+
+# Note: The types OuterProduct resp. InnerProduct are sympy.physics.quantum.operator.OuterProduct/InnerProduct
+#       and by definition are restricted to hold subclasses of KetBase and BraBase. Not to be confused
+#       with other implementations of same or similar mathematical concepts elsewhere in SymPy (e.g.
+#       .matrices or .physics.vector deal with vectors/matrices described by coefficients.
+#       Classical outer/inner product is implemented by sympy.physics.vector.functions.outer resp.
+#       sympy.matrices.dot/sympy.physics.vector.functions.dot).
+
+
+from typing import cast, Any, List, Tuple # keep compatibility with Python 3.8
+from sympy.core.expr import Expr
+from sympy.core.add import Add
+from sympy.core.mul import Mul
+from sympy.core.power import Pow
+from sympy.core.singleton import S
+from sympy.core.numbers import Number
+from sympy.functions.elementary.integers import floor
+
+from sympy.physics.quantum.dagger import Dagger
+
+# Composite types from .quantum that qapply needs to break up or compose:
+from sympy.physics.quantum.anticommutator import AntiCommutator
+from sympy.physics.quantum.commutator import Commutator
+from sympy.physics.quantum.innerproduct import InnerProduct
+from sympy.physics.quantum.operator import OuterProduct
+from sympy.physics.quantum.tensorproduct import TensorProduct
+#from sympy.physics.quantum.density import Density # moved to qapply body
+#otherwise generates circular import reference qapply->Density->represent->qapply
+
+
+
+__all__ = [
+    'qapply'
+]
+
+
+#-----------------------------------------------------------------------------
+# Main code
+#-----------------------------------------------------------------------------
+
+def qapply(e, **options):

```

Comment:
Maybe the keyword arguments should be made explicit here like:
```
def qapply(e, *, dagger=False, ...):
```

Reply:
I usually use these default value style myself. However I hadn't seen it in the original qapply, and have seen the getattr(option, key, default) approach in many places in SymPy. 
In this case however it cannot be used: For several options the default value depends on the value of another option. This can only be handled with the getattr(..) style.
Moreover if there are more than a few options having a clear block of getattr(..) statements in one place - as I have done - is very readable.

Reply:
> This can only be handled with the getattr(..) style.

That just means that it needs to normalise the options like:
```python
def qapply(e, *, option1=False, option2=None):
    if option2 is None:
        option2 = thing1 if option1 else thing2
```
The advantage of specifying the keyword arguments explicitly are:
1. If any argument passed in is misspelled then an error will be seen rather than silently ignored.
2. The options will show up in the docs and under autocomplete in editors.

Also with explicit arguments static analysis tools like mypy can pick up on errors:
```console
$ cat b.py 

def qapply1(e, **options):
    pass

def qapply2(e, *, option1=None, option2=None):
    pass

qapply1(1, bad=False)
qapply2(1, bad=False)

$ mypy b.py
b.py:6: note: "qapply2" defined here
b.py:10: error: Unexpected keyword argument "bad" for "qapply2"  [call-arg]
Found 1 error in 1 file (checked 1 source file)
```

Reply:
Will do the rebase to master when the code base is fixed, as I haven't done this before.

Reply:
Fixed. Used expand() as template.